core/local/steps/overwriting_move.js

```diff
@@ -0,0 +1,145 @@
+/* @flow */
+
+const _ = require('lodash')
+
+const Buffer = require('./buffer')
+const logger = require('../../logger')
+
+/*::
+import type { AtomWatcherEvent, Batch } from './event'
+import type { Metadata } from '../../metadata'
+
+type OverwritingMoveState = {
+  deletedEventsByPath: Map<string, AtomWatcherEvent>,
+  pending: {
+    deletedEventsByPath: Map<string, AtomWatcherEvent>,
+    events: Batch,
+    timeout?: TimeoutID
+  }
+}
+
+type PouchFunctions = {
+  byIdMaybeAsync: (string) => Promise<?Metadata>
+}
+
+type OverwritingMoveOptions = {
+  state: {
+    [typeof STEP_NAME]: OverwritingMoveState
+  }
+}
+*/
+
+const STEP_NAME = 'overwritingMove'
+
+const log = logger({
+  component: `atom/${STEP_NAME}`
+})
+
+// Wait at most this delay (in milliseconds) to fix identical renaming related
+// events.
+const DELAY = 500
+
+const initialState = () => ({
+  [STEP_NAME]: {
+    deletedEventsByPath: new Map/*:: <string,AtomWatcherEvent> */(),
+    pending: {
+      deletedEventsByPath: new Map/*:: <string,AtomWatcherEvent> */(),
+      events: []
+    }
+  }
+})
+
+/** Current batch becomes pending.
+ *
+ * So we can:
+ * - Detect identical renaming events in the next one.
+ * - Possibly aggregate broken deleted events from the current (now pending)
+ *   one.
+ */
+const rotateState = (state, events) => {
+  state.pending = {
+    deletedEventsByPath: state.deletedEventsByPath,
+    events
+  }
+  state.deletedEventsByPath = new Map()
+}
+
+/** Broken deleted event ids match identical renamed ones. */
+const indexDeletedEvent = (event, state) => {
+  if (event.action === 'deleted') {
+    state.deletedEventsByPath.set(event.path, event)
+  }
+}
+
+/** Possibly change event action to 'ignored'.
+ *
+ * In case it is a deleted event related to an overwriting move.
+ */
+const ignoreDeletedBeforeOverwritingMove = (event, state) => {
+  const { path } = event
+  const pendingDeletedEvent = (
+    state.deletedEventsByPath.get(path) ||
+    state.pending.deletedEventsByPath.get(path)
+  )
+  if (pendingDeletedEvent) {
+    event.overwrite = true
+    pendingDeletedEvent.action = 'ignored'
+    _.set(pendingDeletedEvent, [STEP_NAME, 'deletedBeforeRenamed'], event)
```

Comment:
~~`deletedBeforeRenamed`~~ `deletedBy`?
(will read `overwritingMove: {deletedBy: <the move event>}`

Reply:
Actually `deletedBeforeRenamed` is not so bad.